### PR TITLE
Factor out common Buf* implementations into the macro rule.

### PR DESCRIPTION
**Pull request by evgenyzdanovich**
## Description

Factor out common Buf* implementations into the macro rule.

### Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update
-   [x] Refactor
-   [ ] New or updated tests

## Checklist

-   [x] I have performed a self-review of my code.
-   [x] I have commented my code where necessary.
-   [ ] I have updated the documentation if needed.
-   [x] My changes do not introduce new warnings.
-   [x] I have added tests that prove my changes are effective or that my feature works.
-   [x] New and existing tests pass with my changes.

### CODE REVIEW

**Comment by bewakes** [comment]
Just a thought, could using const generics be possible here?

**Comment by evgenyzdanovich** [comment]
> Just a thought, could using const generics be possible here?

@bewakes 

I guess yes, but I can't think of any significant advantages here. Do you have something in mind?

**Comment by bewakes** [comment]
Maybe as an alternative to the macro usage.

**Comment by evgenyzdanovich** [comment]
> Maybe as an alternative to the macro usage.

Ah, I see. But I don't think we'll get rid of macro completely, right? 

Anyways, no strong opinion on this point. @delbonis @storopoli any thoughts?

File: crates/evmexec/src/engine.rs [hunk 1]
```diff
@@ -529,8 +529,8 @@ mod tests {
 
         let timestamp = 0;
         let el_ops = vec![];
-        let safe_l1_block = Buf32(FixedBytes::<32>::random());
-        let prev_l2_block = Buf32(FixedBytes::<32>::random()).into();
+        let safe_l1_block = FixedBytes::<32>::random().into();
+        let prev_l2_block = Buf32(FixedBytes::<32>::random().into()).into();
```
Comment (delbonis) [inline crates/evmexec/src/engine.rs:533]: Could we get rid of this extra `Buf32` here too?

File: crates/primitives/src/buf.rs [hunk 2]
```diff
@@ -1,84 +1,32 @@
...
-    fn from(value: [u8; 32]) -> Self {
-        Self(FixedBytes::from(value))
-    }
-}
-
-impl From<Buf32> for FixedBytes<32> {
-    fn from(value: Buf32) -> Self {
-        value.0
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        hex::decode_to_array(s).map(Self::new)
```
Comment (delbonis) [inline crates/primitives/src/buf.rs:29]: This is great :+1:.

File: crates/primitives/src/macros.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                pub fn as_bytes(&self) -> &[u8] {
+                    self.0.as_slice()
+                }
+
+                pub fn zero() -> Self {
+                    Self([0; $len].into())
+                }
+
+                pub fn is_zero(&self) -> bool {
+                    *self.as_ref() == [0; $len]
```
Comment (delbonis) [inline crates/primitives/src/macros.rs:28]: Is this better than `self.iter().all(|v| == 0)`?

**Review by delbonis — APPROVED** [review]
This is great, appreciate the cleanups downstream.  Just a couple questions though.

**Comment by delbonis** [comment]
@bewakes @evgenyzdanovich It doesn't make a huge improvement and the generics can confuse rust-analyzer.  This is a clean 1:1 generalization, I can't see anywhere we could take advantage of it being generic at currently.

**Comment by storopoli** [comment]
> > Maybe as an alternative to the macro usage.
> 
> Ah, I see. But I don't think we'll get rid of macro completely, right?
> 
> Anyways, no strong opinion on this point. @delbonis @storopoli any thoughts?

I like the idea. Thanks!
Agree with all @delbonis's comments.

File: crates/evmexec/src/engine.rs (hunk 1 above)
Comment (evgenyzdanovich) [reply crates/evmexec/src/engine.rs:533]: we can, but then `L2BlockId` (in the crate `state`) would need an additional `From` impl, I guess.

Given it's in the tests only (I didn't see this pattern in the code, but maybe I overlooked somewhere), I'd leave it as is, if you don't mind.

File: crates/primitives/src/buf.rs (hunk 2 above)
Comment (evgenyzdanovich) [reply crates/primitives/src/buf.rs:29]: thanks :)

**Review by Rajil1213 — APPROVED** [review]
Looks good. No issues with the parts of this PR that touch the bridge crates.

**Review by storopoli — DISMISSED** [review]
ACK 1f0f0280a19f826bb0c2d0cbcd6181318e3a7c5c

**Comment by evgenyzdanovich** [comment]
> Looks good but didn't address all imports. I see `std`, `serde`, `arbitrary` etc not being imported via absolute crate path. `Default` too

Right, done (hopefully).

File: crates/evmexec/src/engine.rs (hunk 1 above)
Comment (delbonis) [reply crates/evmexec/src/engine.rs:533]: I think adding a `From<Buf32>` makes sense there.

Would it make sense add another macro that can instantiate the impls for types that are just wrappers around hashes like `L2BlockId` and `L1BlockId`?

Comment (evgenyzdanovich) [reply crates/evmexec/src/engine.rs:533]: Good idea one another macro, done.

> I think adding a `From<Buf32>` makes sense there.

@delbonis Sorry, did you mean `From<FixedBytes>`? `From<Buf32>` is already there. 
I now think adding `From<FixedBytes>` actually makes little sense. One obvious drawback - it adds a dependency on `reth_pritimitives` for the `strata-state` crate. Right now, `strata-state` depends only on `strata-primitives` which I find ergonomic.

Nonetheless, not a strong opinion, let me know what do you think, so I can wrap this piece of work soon :)

File: crates/primitives/src/macros.rs [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+                fn default() -> Self {
+                    Self([0; $len])
+                }
+            }
+
+            impl ::std::fmt::Debug for $name {
+                fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>) -> ::std::fmt::Result {
+                    // twice as large, required by the hex::encode_to_slice.
+                    let mut buf = [0; $len * 2];
+                    hex::encode_to_slice(self.0, &mut buf).expect("buf: enc hex");
```
[the remark this answers is not in the document]
Comment (evgenyzdanovich) [reply crates/primitives/src/macros.rs]: thanks, done

File: crates/primitives/src/macros.rs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                    f.write_str(unsafe { ::core::str::from_utf8_unchecked(&buf) })
+                }
+            }
+
+            impl ::std::fmt::Display for $name {
+                fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>) -> ::std::fmt::Result {
+                    // fmt only first and last bits of data.
+                    let mut buf = [0; 6];
+                    hex::encode_to_slice(&self.0[..3], &mut buf).expect("buf: enc hex");
```
[the remark this answers is not in the document]
Comment (evgenyzdanovich) [reply crates/primitives/src/macros.rs]: thanks, done

File: crates/primitives/src/macros.rs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                    f.write_str(unsafe { ::core::str::from_utf8_unchecked(&buf) })?;
+                    f.write_str("..")?;
+                    hex::encode_to_slice(&self.0[$len - 3..], &mut buf).expect("buf: enc hex");
```
[the remark this answers is not in the document]
Comment (evgenyzdanovich) [reply crates/primitives/src/macros.rs]: thanks, done

**Review by evgenyzdanovich — COMMENTED** [review]
Thanks for your thorough review, @Zk2u ! PTAL (hopefully final one:)

**Review by prajwolrg — APPROVED** [review]
LGTM!

File: crates/evmexec/src/engine.rs (hunk 1 above)
Comment (delbonis) [reply crates/evmexec/src/engine.rs:533]: Oh then I guess we're all set.  Yeah we want to avoid extra deps on `reth_primitives`.

**Review by delbonis — APPROVED** [review]

**Review by storopoli — APPROVED** [review]
ACK df5234738a2cbbf586fc823925b8bcd3ec51749d

**Pull request merged**